ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs

```diff
@@ -625,12 +641,174 @@ annotateHeader epochSlots =
     splice bs (Right (_leftover, txAux)) =
       (Lazy.toStrict . slice bs) <$> txAux
 
--- Implementation of 'blockHash' derived from
---
--- > blockHashAnnotated :: ABlock ByteString -> HeaderHash
--- > blockHashAnnotated = hashDecoded . fmap wrapHeaderBytes . blockHeader
---
--- I couldn't find a version for headers
-headerHashAnnotated :: CC.Block.AHeader ByteString
-                    -> Crypto.Hash CC.Block.Header
-headerHashAnnotated = Crypto.hashDecoded . fmap CC.Block.wrapHeaderBytes
+{-------------------------------------------------------------------------------
+  Internal auxiliary
+
+  Since we will not be creating further boundary blocks, these utilities do not
+  exist in the cardano-ledger repo, but we need them for the genesis case in the
+  demo.
+  -------------------------------------------------------------------------------}
+
+annotateBoundary :: Crypto.ProtocolMagicId
+                 -> CC.Block.BoundaryValidationData ()
+                 -> CC.Block.BoundaryValidationData ByteString
+annotateBoundary pm =
```

Comment:
While debugging the demo with `--real-pbft` (#705): I discovered that there must be a bug somewhere in this function (or in cardano-ledger#566): if I give it an EBB with `boundaryPrevHash = Left _`, the block I get back after `annotateBoundary` suddenly has `boundaryPrevHash = Right _`. If I replace `annotateBoundary` with `fmap (const mempty)`, then `boundaryPrevHash` remains `Left _`. This is the reason why the first produced block doesn't fit onto the chain.

Can you add a round-trip QC property for `fromCBOR.. . toCBOR..` + find the actual bug? :slightly_smiling_face: